File: Algorithms/0341.nested-interger/nested-interger.go
package _341_nested_interger

// This is the interface that allows for creating nested lists.
// You should not implement it, or speculate about its implementation
type NestedInteger struct {
}

// Return true if this NestedInteger holds a single integer, rather than a nested list.
func (this NestedInteger) IsInteger() bool {
	return false
}

// Return the single integer that this NestedInteger holds, if it holds a single integer
// The result is undefined if this NestedInteger holds a nested list
// So before calling this method, you should have a check
func (this NestedInteger) GetInteger() int {
	return 0
}

// Set this NestedInteger to hold a single integer.
func (n *NestedInteger) SetInteger(value int) {}

// Set this NestedInteger to hold a nested list and adds a nested integer to it.
func (this *NestedInteger) Add(elem NestedInteger) {}

// Return the nested list that this NestedInteger holds, if it holds a nested list
// The list length is zero if this NestedInteger holds a single integer
// You can access NestedInteger's List element directly if you want to modify it
func (this NestedInteger) GetList() []*NestedInteger {
	return []*NestedInteger{}
}

type NestedIterator struct {
	Data []*NestedInteger
}

func Constructor(nestedList []*NestedInteger) *NestedIterator {
	m := &NestedIterator{Data: nestedList}
	return m
}

func (this *NestedIterator) Next() int {
	v := this.Data[0].GetInteger()
	this.Data = this.Data[1:]
	return v
}

func (this *NestedIterator) HasNext() bool {
	if len(this.Data) == 0 {
		return false
	}
	if this.Data[0].IsInteger() {
		return true
	}
	this.Data = append(this.Data[0].GetList(), this.Data[1:]...)
	return this.HasNext()
}

File: Algorithms/0341.nested-interger/nested-interger.py
# """
# This is the interface that allows for creating nested lists.
# You should not implement it, or speculate about its implementation
# """
# class NestedInteger:
#    def isInteger(self) -> bool:
#        """
#        @return True if this NestedInteger holds a single integer, rather than a nested list.
#        """
#
#    def getInteger(self) -> int:
#        """
#        @return the single integer that this NestedInteger holds, if it holds a single integer
#        Return None if this NestedInteger holds a nested list
#        """
#
#    def getList(self) -> [NestedInteger]:
#        """
#        @return the nested list that this NestedInteger holds, if it holds a nested list
#        Return None if this NestedInteger holds a single integer
#        """

from collections import deque


class NestedIterator(object):
    def __init__(self, nestedList):
        self.q = deque()
        for li in nestedList:  # 先把所有的都放到队列中，然后在hasNext方法中依次展开
            self.q.append(li)

    def next(self):
        return self.q.popleft().getInteger()

    def hasNext(self):
        while self.q and self.q[0].isInteger() == False:
            tempLi = self.q.popleft().getList()  # 队列中第一个是列表，把他弹出并展开在放到队列里
            for i in tempLi[::-1]:  # 倒序放
                self.q.appendleft(i)  # 从队列头倒着放
        return len(self.q) > 0

# Your NestedIterator object will be instantiated and called as such:
# i, v = NestedIterator(nestedList), []
# while i.hasNext(): v.append(i.next())
